File: launcher.cs
// ──────────────────────────────────────────────────────────────────────────
//  JDrakoon3 launcher (C# / winexe) — RUNTIME ONLY.
//
//  Compiled with csc /target:winexe so it is a true GUI-subsystem app: it
//  NEVER shows a console window. Icon is embedded at compile time
//  (/win32icon) — no rcedit, no PE patching, nothing to corrupt.
//
//  Everything is prebuilt by build-release.ps1. At runtime we only:
//    1. free port 3001,
//    2. start the prebuilt backend with the bundled node.exe (hidden),
//    3. wait for it to answer, then open Edge in kiosk mode,
//    4. shut down when EITHER the kiosk closes OR the backend exits
//       (e.g. the in-app Shutdown button).
//
//  No git, no npm, no compiling at startup. Users need only .NET Framework 4
//  (built into Windows 10/11).
// ──────────────────────────────────────────────────────────────────────────
using System;
using System.Diagnostics;
using System.IO;
using System.Net;
using System.Threading;

[assembly: System.Reflection.AssemblyTitle("JDrakoon3")]
[assembly: System.Reflection.AssemblyProduct("JDrakoon3")]
[assembly: System.Reflection.AssemblyCompany("Lonezsi")]
[assembly: System.Reflection.AssemblyFileVersion("3.0.1")]

static class Launcher
{
    const int PORT = 3001;
    const string URL = "http://localhost:3001";

    static string ROOT, BACKEND_DIR, DATA_DIR, LOG, BACKEND_LOG, EDGE_PROFILE;
    static Process backend, kiosk;
    static readonly object ioLock = new object();

    static void Log(string msg)
    {
        try
        {
            lock (ioLock)
                File.AppendAllText(LOG, "[" + DateTime.Now.ToString("s") + "] " + msg + "\r\n");
        }
        catch { }
    }

    static int Main()
    {
        ROOT = AppDomain.CurrentDomain.BaseDirectory.TrimEnd('\\');
        BACKEND_DIR = Path.Combine(ROOT, "backend");
        string local = Environment.GetEnvironmentVariable("LOCALAPPDATA");
        if (string.IsNullOrEmpty(local)) local = Path.GetTempPath();
        DATA_DIR = Path.Combine(local, "JDrakoon3");
        try { Directory.CreateDirectory(DATA_DIR); } catch { }
        LOG = Path.Combine(DATA_DIR, "launcher.log");
        BACKEND_LOG = Path.Combine(DATA_DIR, "backend.log");
        EDGE_PROFILE = Path.Combine(DATA_DIR, "edge-profile");

        try
        {
            Log("Launching from " + ROOT);
            if (!File.Exists(Path.Combine(BACKEND_DIR, "dist", "index.js")))
            {
                Log("FATAL: backend/dist/index.js missing - incomplete build.");
                return 1;
            }

            KillPort();
            StartBackend();

            if (!WaitForBackend())
            {
                Log("Backend never answered.");
                Cleanup();
                return 1;
            }

            LaunchKiosk();

            // Block until the kiosk window closes OR the backend exits.
            while (true)
            {
                Thread.Sleep(400);
                if (HasExited(kiosk)) { Log("Kiosk closed."); break; }
                if (HasExited(backend)) { Log("Backend exited."); break; }
            }

            Cleanup();
            return 0;
        }
        catch (Exception ex)
        {
            Log("FATAL: " + ex.Message);
            Cleanup();
            return 1;
        }
    }

    static bool HasExited(Process p)
    {
        try { return p != null && p.HasExited; } catch { return false; }
    }

    static string NodeBin()
    {
        string bundled = Path.Combine(ROOT, "node.exe");
        return File.Exists(bundled) ? bundled : "node";
    }

    static void StartBackend()
    {
        string node = NodeBin();
        Log("Starting backend with " + node);
        try { File.WriteAllText(BACKEND_LOG, ""); } catch { }

        var psi = new ProcessStartInfo
        {
            FileName = node,
            Arguments = "dist/index.js",
            WorkingDirectory = BACKEND_DIR,
            UseShellExecute = false,
            CreateNoWindow = true,
            RedirectStandardOutput = true,
            RedirectStandardError = true,
        };
        psi.EnvironmentVariables["NODE_ENV"] = "production";

        backend = new Process { StartInfo = psi, EnableRaisingEvents = true };
        DataReceivedEventHandler sink = (s, e) =>
        {
            if (e.Data == null) return;
            try { lock (ioLock) File.AppendAllText(BACKEND_LOG, e.Data + "\r\n"); } catch { }
        };
        backend.OutputDataReceived += sink;
        backend.ErrorDataReceived += sink;
        backend.Start();
        backend.BeginOutputReadLine();
        backend.BeginErrorReadLine();
    }

    static bool WaitForBackend()
    {
        for (int i = 0; i < 60; i++)
        {
            try
            {
                var req = (HttpWebRequest)WebRequest.Create(URL + "/api/version");
                req.Timeout = 1500;
                using (var resp = (HttpWebResponse)req.GetResponse())
                    if ((int)resp.StatusCode < 500) return true;
            }
            catch { }

            if (HasExited(backend)) return false; // crashed on startup
            Thread.Sleep(500);
        }
        return false;
    }

    static string FindEdge()
    {
        string[] roots =
        {
            Environment.GetFolderPath(Environment.SpecialFolder.ProgramFilesX86),
            Environment.GetFolderPath(Environment.SpecialFolder.ProgramFiles),
        };
        foreach (var r in roots)
        {
            if (string.IsNullOrEmpty(r)) continue;
            string p = Path.Combine(r, "Microsoft\\Edge\\Application\\msedge.exe");
            if (File.Exists(p)) return p;
        }
        return null;
    }

    static void LaunchKiosk()
    {
        string edge = FindEdge();
        if (edge == null)
        {
            Log("Edge not found - opening default browser (no kiosk).");
            try { Process.Start(new ProcessStartInfo { FileName = URL, UseShellExecute = true }); }
            catch (Exception e) { Log("default browser failed: " + e.Message); }
            return;
        }

        // A dedicated user-data-dir forces a fresh, isolated Edge instance whose
        // process we can reliably watch (otherwise --kiosk may attach to an
        // already-running Edge and our handle exits instantly).
        string args =
            "--kiosk " + URL +
            " --edge-kiosk-type=fullscreen --no-first-run --no-default-browser-check" +
            " --user-data-dir=\"" + EDGE_PROFILE + "\"";

        Log("Opening kiosk.");
        kiosk = Process.Start(new ProcessStartInfo
        {
            FileName = edge,
            Arguments = args,
            UseShellExecute = false,
        });
    }

    static void Run(string file, string args, bool wait)
    {
        try
        {
            var p = Process.Start(new ProcessStartInfo
            {
                FileName = file,
                Arguments = args,
                UseShellExecute = false,
                CreateNoWindow = true,
                RedirectStandardOutput = true,
            });
            if (wait) { p.StandardOutput.ReadToEnd(); p.WaitForExit(); }
        }
        catch { }
    }

    static void KillPort()
    {
        try
        {
            var psi = new ProcessStartInfo
            {
                FileName = "netstat.exe",
                Arguments = "-ano",
                UseShellExecute = false,
                CreateNoWindow = true,
                RedirectStandardOutput = true,
            };
            var p = Process.Start(psi);
            string outp = p.StandardOutput.ReadToEnd();
            p.WaitForExit();

            foreach (var line in outp.Split('\n'))
            {
                if (line.IndexOf("LISTENING", StringComparison.OrdinalIgnoreCase) < 0) continue;
                var parts = line.Trim().Split(new[] { ' ', '\t' }, StringSplitOptions.RemoveEmptyEntries);

                bool onPort = false;
                foreach (var tok in parts)
                    if (tok.EndsWith(":" + PORT, StringComparison.Ordinal)) { onPort = true; break; }
                if (!onPort) continue;

                int pid;
                if (parts.Length >= 1 && int.TryParse(parts[parts.Length - 1], out pid) && pid > 0)
                {
                    Run("taskkill.exe", "/PID " + pid + " /F", true);
                    Log("Freed port " + PORT + " (killed pid " + pid + ")");
                }
            }
        }
        catch { }
    }

    static void Cleanup()
    {
        Log("Shutting down.");
        try { if (kiosk != null && !kiosk.HasExited) Run("taskkill.exe", "/PID " + kiosk.Id + " /F /T", true); } catch { }
        try { if (backend != null && !backend.HasExited) backend.Kill(); } catch { }
        KillPort();
        Log("Done.");
    }
}

File: launcher.mjs
// ──────────────────────────────────────────────────────────────────────────
//  JDrakoon3 launcher — macOS / Linux (the cross-platform counterpart to
//  launcher.cs, which stays the Windows path with its WebView2 host).
//
//  Pure Node, zero deps. Mirrors the proven Windows lifecycle:
//    1. free port 3001 (best-effort),
//    2. start the prebuilt backend with THIS node (hidden, logged),
//    3. wait for it to answer,
//    4. open a Chromium-family browser in --kiosk fullscreen (own profile),
//    5. shut down when EITHER the kiosk closes OR the backend exits.
//
//  Run:  node launcher.mjs        (needs Node 18+ and a Chromium browser;
//                                   falls back to the default browser, no kiosk)
// ──────────────────────────────────────────────────────────────────────────
import { spawn, execSync } from "node:child_process";
import { existsSync, mkdirSync, appendFileSync, openSync } from "node:fs";
import { homedir } from "node:os";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";
import http from "node:http";

const PORT = 3001;
const BASEURL = `http://127.0.0.1:${PORT}`;
const ROOT = dirname(fileURLToPath(import.meta.url));
const BACKEND_DIR = join(ROOT, "backend");
const IS_MAC = process.platform === "darwin";

const DATA_DIR = join(homedir(), ".jdrakoon3");
mkdirSync(DATA_DIR, { recursive: true });
const LOG = join(DATA_DIR, "launcher.log");
const BACKEND_LOG = join(DATA_DIR, "backend.log");
const PROFILE = join(DATA_DIR, "browser-profile");

const log = (m) => {
  const line = `[${new Date().toISOString()}] ${m}\n`;
  try { appendFileSync(LOG, line); } catch {}
  process.stdout.write(line);
};

let backend = null;
let kiosk = null;
let shuttingDown = false;

// ── free the port (best-effort; the backend would fail to bind otherwise) ──
function killPort() {
  try {
    // lsof is on macOS and most Linux; ignore if absent.
    const pids = execSync(`lsof -ti tcp:${PORT}`, { stdio: ["ignore", "pipe", "ignore"] })
      .toString()
      .split(/\s+/)
      .filter(Boolean);
    for (const pid of pids) {
      try { process.kill(Number(pid), "SIGKILL"); log(`Freed port ${PORT} (pid ${pid})`); } catch {}
    }
  } catch { /* nothing listening, or lsof missing */ }
}

// ── start the backend with the same node that's running this launcher ──────
function startBackend() {
  const entry = join(BACKEND_DIR, "dist", "index.js");
  if (!existsSync(entry)) {
    log(`FATAL: ${entry} missing — run the build first.`);
    process.exit(1);
  }
  const out = openSync(BACKEND_LOG, "w");
  log(`Starting backend with ${process.execPath}`);
  backend = spawn(process.execPath, ["dist/index.js"], {
    cwd: BACKEND_DIR,
    env: { ...process.env, NODE_ENV: "production" },
    stdio: ["ignore", out, out],
  });
  backend.on("exit", (code) => {
    log(`Backend exited (${code}).`);
    if (!shuttingDown) cleanup(0);
  });
}

// ── poll until the backend answers ─────────────────────────────────────────
function waitForBackend(tries = 60) {
  return new Promise((resolve) => {
    const attempt = (n) => {
      if (n <= 0) return resolve(false);
      const req = http.get(`${BASEURL}/api/version`, (res) => {
        res.resume();
        res.statusCode < 500 ? resolve(true) : retry(n);
      });
      req.on("error", () => retry(n));
      req.setTimeout(1500, () => { req.destroy(); retry(n); });
    };
    const retry = (n) => setTimeout(() => attempt(n - 1), 500);
    attempt(tries);
  });
}

// ── find a Chromium-family browser for true kiosk mode ─────────────────────
function findBrowser() {
  if (IS_MAC) {
    const apps = [
      "/Applications/Google Chrome.app/Contents/MacOS/Google Chrome",
      "/Applications/Microsoft Edge.app/Contents/MacOS/Microsoft Edge",
      "/Applications/Brave Browser.app/Contents/MacOS/Brave Browser",
      "/Applications/Chromium.app/Contents/MacOS/Chromium",
    ];
    return apps.find((p) => existsSync(p)) || null;
  }
  for (const cmd of [
    "google-chrome", "google-chrome-stable", "chromium", "chromium-browser",
    "microsoft-edge", "microsoft-edge-stable", "brave-browser",
  ]) {
    try {
      const p = execSync(`command -v ${cmd}`, { stdio: ["ignore", "pipe", "ignore"] }).toString().trim();
      if (p) return p;
    } catch {}
  }
  return null;
}

function launchKiosk() {
  const browser = findBrowser();
  if (!browser) {
    log("No Chromium browser found — opening default browser (no kiosk).");
    const opener = IS_MAC ? "open" : "xdg-open";
    try { spawn(opener, [BASEURL], { detached: true, stdio: "ignore" }).unref(); }
    catch (e) { log(`default browser failed: ${e.message}`); }
    return;
  }
  log(`Opening kiosk: ${browser}`);
  kiosk = spawn(
    browser,
    [
      "--kiosk", BASEURL,
      "--no-first-run", "--no-default-browser-check",
      "--disable-session-crashed-bubble", "--noerrdialogs",
      `--user-data-dir=${PROFILE}`,
    ],
    { stdio: "ignore" },
  );
  kiosk.on("exit", () => { log("Kiosk closed."); if (!shuttingDown) cleanup(0); });
}

function cleanup(code) {
  if (shuttingDown) return;
  shuttingDown = true;
  log("Shutting down.");
  try { if (kiosk && kiosk.exitCode === null) kiosk.kill("SIGTERM"); } catch {}
  try { if (backend && backend.exitCode === null) backend.kill("SIGTERM"); } catch {}
  killPort();
  log("Done.");
  process.exit(code);
}

process.on("SIGINT", () => cleanup(0));
process.on("SIGTERM", () => cleanup(0));

// ── main ────────────────────────────────────────────────────────────────────
log(`Launching from ${ROOT}`);
killPort();
startBackend();
const up = await waitForBackend();
if (!up) {
  log("Backend never answered.");
  cleanup(1);
} else {
  launchKiosk();
}
